### PR TITLE
Team UID From Int to String

### PR DESCRIPTION
**Pull request by ZhangCreations**
## Description

**Merge this after #34**

- Changed the team IDs type from ints to 6 char string in Sequelize model so that in the future, custom storefront URLs can be implemented.
- Changed team ID type everywhere else it appears

Link to ticket: https://www.notion.so/uwblueprintexecs/Change-team-UID-to-a-4-character-format-35e2fd5749c64d2a83075be4030e638e

## Approach

The approach here was to use other people's available work and then put it into our code. Using the [nanoid-good library](https://github.com/y-gagar1n/nanoid-good), to generate the 6 char values with the bonus of ensuring no profane English words! When a team is created, the id is created and attempts to enter that Team entry into the Sequelize Team model. If the id is already used, then the `unique: true` line in 'teams.js' ensures that a `SequelizeUniqueConstraintError` will be thrown and when that happens, it'll just re-attempt until a unique key is generated (more accurately 999 times but assuming small enough user base a unique id will be generated before that limit is reached).

## Extra Fun Math that I Want to Include for Completeness

The alphabet for id generation consists of `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ` (because strings with these combinations are easy to remember). With 36 characters in that string, assuming 1000 teams, the probability of a collision is 4.593 * 10<sup>-7</sup>%.

Basically **collision on generation is super rare and when it does the penalty is minimal** (If it happens to you, buy lottery tickets)

## Testing

Testing was done in the graphql playground on all team queries and mutations. Collisions were tested by reducing the alphabet and string size for the nanoid. If you have previously set up the db and want to test out this code, you'll need to reset the migration with the following steps

1. `docker-compose up --build -d`
2. `docker ps`
3. `docker exec -it <container_id> /bin/bash` from there run the following commands
5. `sequelize db:migrate:undo:all`
6. `sequelize db:migrate`

### CODE REVIEW

**Comment by ZhangCreations** [comment]
Closed due to a major issue found in retrying upon duplicate, opening back up again because it has been fixed.

File: src/models/team.js [hunk 1]
```diff
@@ -1,4 +1,10 @@
 const { Model } = require('sequelize');
+var en = require('nanoid-good/locale/en');
+var customAlphabet = require('nanoid-good').customAlphabet(en);
+
+// Based on documentation here https://github.com/ai/nanoid
+const alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
```
Comment (jzxhuang) [inline src/models/team.js]: i think we should keep it to letters only, thoughts?

File: src/models/team.js [hunk 2]
```diff
@@ -13,10 +19,33 @@ module.exports = (sequelize, DataTypes) => {
...
-      name: DataTypes.STRING,
-      organization: DataTypes.STRING,
-      amountRaised: DataTypes.DECIMAL(20, 2),
-      itemsSold: DataTypes.INTEGER,
+      id: {
+        type: DataTypes.STRING,
+        allowNull: false,
+        primaryKey: true,
+        unique: true,
+        defaultValue: generator,
```
Comment (jzxhuang) [inline src/models/team.js:29]: neat

File: src/resolvers/teams.js [hunk 3]
```diff
@@ -42,12 +44,27 @@ const teamsResolvers = {
...
+              },
+              {
+                transaction: t,
+              },
+            );
+          });
+        },
+        {
+          max: 999,
+          match: [Sequelize.UniqueConstraintError],
```
Comment (jzxhuang) [inline src/resolvers/teams.js:64]: this is kinda cool

**Review by jzxhuang — COMMENTED** [review]
looking gooddddd minor comments

nanoid-good seems like a cool library

File: src/models/team.js (hunk 1 above)
Comment (ZhangCreations) [reply src/models/team.js]: Shouldn't be an issue collision-wise, as long as the user base is small enough the probabilities are still in our favour. I'll make this change.

File: src/models/team.js [hunk 4]
```diff
@@ -1,4 +1,11 @@
 const { Model } = require('sequelize');
+var en = require('nanoid-good/locale/en');
+var fr = require('nanoid-good/locale/fr');
+var customAlphabet = require('nanoid-good').customAlphabet(en, fr);
+
+// Based on documentation here https://github.com/ai/nanoid
+const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
```
Comment (ZhangCreations) [inline src/models/team.js:7]: Altered the alphabet according to the previous comment and also added support for french swear word blocking.

**Review by jzxhuang — APPROVED** [review]
seems like it shoudl work

the math you included in the PR description is pretty fun 

btw I noticed you playing around with `package-lock.json`, you should generally commit that file (any changes you get from running `npm i` should be committed)

merge after #34 like you said
